packages/frontend/src/hooks/useSearchedResolvedKPITokens.ts

```diff
@@ -0,0 +1,116 @@
+    const specification = token.specification;
+
+    return (
+        includesQuery(specification.title, query) ||
+        includesQuery(specification.description, query) ||
+        includesQuery(specification.tags, query)
+    );
+};
+
+const includesQuery = (value: string | string[], query: string) => {
+    const result =
+        typeof value === "string"
+            ? value.toString().toLowerCase()
+            : value.map((i) => i.toLowerCase());
+    return result.includes(query);
+};
+
+export function useSearchedResolvedKPITokens(searchQuery: string) {
+    const { chain } = useNetwork();
+    const { loading: loadingKPITokens, kpiTokens } = useKPITokens();
+    const ipfsGatewayURL = useIPFSGatewayURL();
+
+    const [loading, setLoading] = useState(false);
+    const [searchableKPITokens, setSearchableKPITokens] = useState<
+        ResolvedKPIToken[]
+    >([]);
+    const [debouncedSearchableKPITokens, setDebouncedSearchableKPITokens] =
+        useState<ResolvedKPIToken[]>([]);
+    const [searchedKPITokens, setSearchedKPITokens] = useState<
+        ResolvedKPIToken[]
+    >([]);
+
+    useDebounce(
+        () => {
+            setDebouncedSearchableKPITokens(searchableKPITokens);
+        },
+        300,
+        [searchableKPITokens]
+    );
+
+    useEffect(() => {
+        setSearchableKPITokens([]);
+        setDebouncedSearchableKPITokens([]);
+        setSearchedKPITokens([]);
+    }, [chain?.id]);
+
+    useEffect(() => {
+        let cancelled = false;
+        const kpiTokensArray = Object.values(kpiTokens);
+        if (!kpiTokens || kpiTokensArray.length === 0) return;
+        const fetchData = async () => {
+            if (!cancelled) setLoading(true);
+            const promises = kpiTokensArray.map(async (kpiToken) => {
+                return (
+                    await Fetcher.resolveKPITokens({
+                        ipfsGatewayURL,
+                        kpiTokens: [kpiToken],
+                    })
+                )[kpiToken.address];
+            });
+            for (const promise of promises) {
+                promise
+                    .then((result) => {
+                        setSearchableKPITokens((prevState) => [
+                            ...prevState,
+                            result,
+                        ]);
+                    })
+                    .catch((error) => {
+                        console.warn(
+                            "error while fetching searchable resolved kpi tokens",
+                            error
+                        );
+                    });
+            }
+            await Promise.allSettled(promises);
+            if (!cancelled) setLoading(false);
+        };
+        void fetchData();
+        return () => {
+            cancelled = true;
+        };
+    }, [ipfsGatewayURL, kpiTokens]);
+
+    useEffect(() => {
+        if (!searchQuery) setSearchedKPITokens(debouncedSearchableKPITokens);
```

Comment:
We should add an `else` here, or `return` inside the `if` statement, otherwise, the second `setSearchedKPITokens` will always run and overwrite the first one.

Reply:
Addressed in d65f11899d21544c1aa81bf671bd18463e535e6c